plugins/pulp_python/plugins/models.py

```diff
@@ -1,120 +1,150 @@
-                        'following attributes are all present: %(attrs)s')
-                msg = msg % {
-                    'attrs': ', '.join([cls._metadata_label(attr) for attr in REQUIRED_ATTRS])}
-                raise ValueError(msg)
-
-            # Add the filename, checksum, and checksum_type to the attrs
-            attrs['_filename'] = '%s-%s.tar%s' % (attrs['name'], attrs['version'], compression_type)
-            attrs['_checksum'] = checksum
-            attrs['_checksum_type'] = DEFAULT_CHECKSUM_TYPE
-            attrs['_metadata_file'] = metadata_file_name
-            package = cls(**attrs)
-            return package
-        finally:
-            if 'package_archive' in locals():
-                package_archive.close()
+    def from_json(cls, package_data, release, project_data):
+        """
+        Create and return (but do not save) an instance of a single Python package object from the
+        metadata available from PyPI JSON.
+
+        :param package_data: metadata specific to a file
+        :type  package_data: dict
+        :param release: version number
+        :type  release: basestring
+        :param project_data: metadata that applies to all versions
+        :type  project_data: dict
+        :return: instance of a package
+        :rtype:  pulp_python.plugins.models.Package
+        """
+        package_attrs = {}
+
+        package_attrs['version'] = release
+        package_attrs['name'] = project_data['name']
+        package_attrs['author'] = project_data['author']
+        package_attrs['summary'] = project_data['summary']
+
+        package_attrs['filename'] = package_data['filename']
+        package_attrs['path'] = package_data['path']
+        package_attrs['packagetype'] = package_data['packagetype']
+        package_attrs['md5_digest'] = package_data['md5_digest']
+
+        # If we are syncing from PyPI, there will be no `checksum`, but will be `md5_digest`
+        package_attrs['_checksum'] = package_data.get('checksum', package_attrs['md5_digest'])
+        package_attrs['_checksum_type'] = package_data.get('checksum_type', 'md5')
+        return cls(**package_attrs)
+
+    @classmethod
+    def from_archive(cls, path):
+        """
+        Create and return (but do not save) an instance of a Python package object from the package
+        itself.
+
+        Twine is smart enough to crack open a tarball, zip, or wheel and parse the metadata that is
+        contained in the package.
+
+        :param path: path to the package
+        :type  path: basestring
+        :return: instance of a package
+        :rtype:  pulp_python.plugins.models.Package
+        """
+        meta_dict = PackageFile.from_filename(path, comment='').metadata_dictionary()
+        filtered_dict = {}
+        # Use only a subset of the attributes in the metadata to keep package minimal.
+        for key, value in meta_dict.iteritems():
+            if key in PACKAGE_ATTRS:
```

Comment:
I see ``md5_digest`` isn't in ``PACKAGE_ATTRS``. I actually think that's probably for the best since it may be a different digest type in the future, but I wonder if we want to make sure at least one checksum is provided by the metadata and use it.